distributed/scheduler.py

```diff
@@ -782,18 +782,19 @@ def _remove_from_task_prefix_count(self, ts: TaskState) -> None:
     def remove_replica(self, ts: TaskState) -> None:
         """The worker no longer has a task in memory"""
         if self.scheduler.validate:
-            assert self in ts.who_has
+            assert self in (ts.who_has or ())
             assert ts in self.has_what
             assert ts not in self.needs_what
 
         self.nbytes -= ts.get_nbytes()
         del self._has_what[ts]
-        ts.who_has.remove(self)
+        if ts.who_has:
+            ts.who_has.remove(self)
```

Comment:
The fact that this was a `remove` and not a discard already proves that this is a non-empty set
```suggestion
        ts.who_has.remove(self)
```

Reply:
one could consider something like

```python
ts.who_has.remove(self)
if not ts.who_has:
    ts.who_has = None
```
but I'm not sure if that's worth it

Reply:
https://github.com/dask/distributed/pull/8331/commits/8cd2b9e42374bfdc81d3b6fa2b34c0df4d6f0722